Adding Fluid file format README. (#533)
fltk/fltk: fluid/README_fd.txt

Note: As there are no start and end markers to a string, a reader must know
      when these strings appear and be prepared to read them correctly, even if
      the String itself is discared after reading.

Content
=======

Header
------

The header for any .fd file is

  # data file for the Fltk User Interface Designer (fluid)

followed by a newline, followed by

  version <float v>

wehere 'v' is the version number as in FL_VERSION (major*1.0 + minor * 0.01
+ patch * 0.0001).

Note: the version number corresponds not so much to the version of Fluid, but
      to the version of the underlying fltk library. So unless the version of
      fltk is finalised, the file format in the github master branch can still
      change unexpectedly.

Note: if the version number is above the internal version number, Fluid will
      abort reading the file. There are no other uses inside the Fluid reader
      except for fltk2 which is beyond the scope of thsi document.

Note: fdesign files start with the text "Magic:". Fluid can read thse files,
      but Forms/XFroms file are beyond the scope of this document.

Options
-------

> Options are usually composed of a Word, two Words, or a Word and a String. If

Reviewer: I think "comprised" is closer to the intended meaning

  "linux_shell_flags" <int> :  : as above for Linux

  "mac_shell_cmd" <str> : as above for macOS

  "mac_shell_flags" <int> : as above for macOS

Note: There is no keyword that marks the ending of the options section. The
      Tree section starts when a word is not in the options list. If the word
      is not a vaild type, Fluid will give an error message and try to continue
      to read the file.

Tree
====

If a keyword is read that is not in the option list, we start reading types.
Types represent all possible entries in the hierarchy including C functions,
class definitions, and of course all widgets. A type is any of the supported
widgets class names, or one of the following:

Function, code, codeblock, decl, data, declblock, comment, class, widget_class

Every type keyword is followed by a word, which is usually interpreted as the
C++ name of the Type, followed by an opening `{`, a list of properties, and
a closing ‘}’. If the type has children, they are stored between another
opening ‘{‘, followed by a list of types, followed by a closing ‘}’.

The file end when there are no more types.

Note: the "class" type has an additional Word following immediatley after

>       the keyword conatining the prefix for the class. So a class definition may

Reviewer: ```suggestion
      the keyword containing the prefix for the class. So a class definition may
```

   |    |    +-- Fl_Tabs
   |    |    +-- Fl_Scroll
   |    |    +-- Fl_Tile
   |    |    +-- Fl_Wizard
   |    +-- Fl_Menu_Type  (note: can't be written)
   |    |    +-- Fl_Menu_Button
   |    |    +-- Fl_Choice
   |    |    +-- Fl_Input_Choice
   |    |    +-- Fl_Menu_Bar
   |    |    +-- Fl_
   |    +-- Fl_Box
   |    +-- Fl_Button
   |    |    +-- Fl_Return_Button
   |    |    +-- Fl_Light_Button
   |    |    +-- Fl_Check_Button
   |    |    +-- Fl_Round_Button
   |    +-- Fl_Repeat_Button
   |    +-- Fl_Browser
   |    +-- Fl_Check_Browser
   |    +-- Fl_Tree
   |    +-- Fl_File_Browser
   |    +-- Fl_Counter
   |    +-- Fl_Spinner
   |    +-- Fl_Input
   |    |    +-- Fl_Output
   |    +-- Fl_File_Input
   |    +-- Fl_Text_Display
   |    +-- Fl_Text_Editor
   |    |    +-- Fl_Simple_Terminal
   |    +-- Fl_Clock
   |    +-- Fl_Help_View
   |    +-- Fl_Progress
   |    +-- Fl_Adjuster
   |    +-- Fl_Dial
   |    +-- Fl_Roller
   |    +-- Fl_Slider
   |    |    +-- Fl_Scrollbar
   |    |    +-- Fl_Value_Slider
   |    +-- Fl_Value_Input
   |    +-- Fl_Value_Output
   :
   :

Properties
----------

Properties are defined in a quite arbitrary manner and on a per-type basis.

> They can be inherited from super types, or be limited by 'is_some_tyep()'

Reviewer: ```suggestion
They can be inherited from super types, or be limited by 'is_some_type()'
```

calls. All properties are optional, some are mutually exclusive.

All Types <word>

  “label” <word> : text
  “user_data” <word> : a value or an expression
  “user_data_type” <word> : usually “void*” or “long”
  “callback” <word> : a function name or a function body
  “comment” <word> : one or many lines of text

>   “open” : group centent visible in the Fluid tree browser

Reviewer: ```suggestion
  “open” : group content visible in the Fluid tree browser
```

  “selected” : type was selected in tree view

Type "Function" <word> : function signature

  none or "private" or "protected" : for methods in classes, or to mark
      functions static in a file
  “C” : if set, function is extern “C”
  “return_type” <word> : C or C++ type descriptor, can start with “virtual” and/or “static”
  ... : inherit more from Fl_Type

Type codeblock <word> : C++ code, usually an 'if

  "after" <word> : C++ code or comment following the closing '}'
  ... : inherit more from Fl_Type

> Type "decl" <word> : C++ code to decalre a variable or cllass member

Reviewer: ```suggestion
Type "decl" <word> : C++ code to declare a variable or class member
```

  none or "public" or "private" or "protected" : for declarations within classes
      defaults to "prvate"

>   none or "local" or "global": for declarition in the code body

Reviewer: ```suggestion
  none or "local" or "global": for declarations in the code body
```

      defaults to "global"
  ... : inherit more from Fl_Type

Type "data" <word> : C++ variable name

  "filename" <word> : name or path as entered by user, forward slashes
  "textmode" : defaults to binary mode
  ... : inherit more from decl

Type "declblock" <word> : C++ code

  none or "public" or "protected" : defaults to private
  "after" <word> : C++ code or comment following the block
  ... : inherit more from Fl_Type

Type "comment" <word> : comment text

  "in_source" or "not_in_source": default to in_source
  "in_header" or "not_in_header": default to in_header
  ... : inherit more from Fl_Type

Type "class" <word> <word> : prefix, class name

  none or "private" or "protected" : defaults to public
  ":" <word> : name of super class
  ... : inherit more from Fl_Type

Type "Fl_Widget" <word> : C++ variable name

  none or "private" or "protected" : ...
  "xywh" <word> : "{%d %d %d %d}" ...
  "tooltip" <word> : tooltip text
  "image" <word> : image name
  "deimage" <word> : deactivated image name
  "type" <word> : integer
  "box" <word> : text or integer (see fltk boxtypes)

>   "down_box" <word> : (is_botton() or Fl_Input_choice" or is_menu_button())

Reviewer: ```suggestion
  "down_box" <word> : (is_button() or Fl_Input_choice" or is_menu_button())
```

      text or integer (see fltk boxtypes)

>   "value" <word> : (is_burron()) integer

Reviewer: ```suggestion
  "value" <word> : (is_button()) integer
```

  "value" <word> : (is_valuator(), is_spinner()) double
  "color" <word> :
      If word starts with "0x", the rest of the field is a hex number.

>       If two intergers follow, this is color and selection_color (not written).

Reviewer: ```suggestion
      If two integers follow, this is color and selection_color (not written).
```
Missed this one

      If one integer follows, it's the color index.
  "selection_color" <word> : integer color index
  "labeltype" <word> :
      If the word is "image", TBD.
      Or one of "NORMAL_LABEL", "SHADOW_LABEL", "ENGRAVED_LABEL",
      "EMBOSSED_LABEL", or "NO_LABEL"

>   "labelfont" <word> : intger

Reviewer: ```suggestion
  "labelfont" <word> : integer
```
Same typo is repeated several times below
